File: src/plugins/botinfo.ts
import { IMatrixClient, Plugin, CommandDef, MessageContext } from "./base";
import { getDb } from "../db";
import fs from "fs";
import path from "path";
import logger from "../utils/logger";

const LLM_ENABLED = (process.env.OLLAMA_HOST ?? "") !== "" && (process.env.OLLAMA_MODEL ?? "") !== "";
const OLLAMA_HOST = (() => {
  const raw = process.env.OLLAMA_HOST ?? "";
  return raw && !raw.startsWith("http") ? `http://${raw}` : raw;
})();
const startTime = Date.now();
let messagesProcessed = 0;

export function incrementMessageCount(): void {
  messagesProcessed++;
}

export class BotInfoPlugin extends Plugin {
  constructor(client: IMatrixClient) {
    super(client);
  }

  get name() {
    return "botinfo";
  }

  get commands(): CommandDef[] {
    return [
      { name: "botinfo", description: "Bot diagnostics", adminOnly: true },
    ];
  }

  async onMessage(ctx: MessageContext): Promise<void> {
    incrementMessageCount();

    if (this.isCommand(ctx.body, "botinfo")) {
      await this.handleBotInfo(ctx);
    }
  }

  private async handleBotInfo(ctx: MessageContext): Promise<void> {
    if (!this.isAdmin(ctx.sender)) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Admin only.");
      return;
    }

    const db = getDb();
    const uptimeMs = Date.now() - startTime;
    const uptime = this.formatUptime(uptimeMs);

    // DB size
    const dataDir = process.env.DATA_DIR ?? "./data";
    const dbPath = path.join(dataDir, "freebee.db");
    let dbSize = "unknown";
    try {
      const stats = fs.statSync(dbPath);
      dbSize = this.formatBytes(stats.size);
    } catch { /* ignore */ }

    // Active reminders
    let activeReminders = 0;
    try {
      const row = db.prepare(`SELECT COUNT(*) as count FROM reminders WHERE fired = 0`).get() as { count: number };
      activeReminders = row.count;
    } catch { /* ignore */ }

    // Total room messages
    let totalRoomMessages = 0;
    try {
      const row = db.prepare(`SELECT SUM(total_messages) as total FROM room_milestones`).get() as { total: number | null };
      totalRoomMessages = row?.total ?? 0;
    } catch { /* ignore */ }

    // LLM status
    let llmStatus = "disabled";
    if (LLM_ENABLED) {
      try {
        const res = await fetch(`${OLLAMA_HOST}/api/tags`, { signal: AbortSignal.timeout(5000) });
        llmStatus = res.ok ? `running (${process.env.OLLAMA_MODEL})` : `error (HTTP ${res.status})`;
      } catch {
        llmStatus = "error (unreachable)";
      }
    }

    // Estimated LLM tokens (rough: ~1.3 tokens per word for English)
    let estimatedTokens = "N/A";
    if (LLM_ENABLED) {
      try {
        const row = db.prepare(`SELECT COUNT(*) as classified FROM llm_classifications`).get() as { classified: number };
        const wordRow = db.prepare(`SELECT SUM(total_words) as words FROM user_stats`).get() as { words: number | null };
        const totalWords = wordRow?.words ?? 0;
        const tokens = Math.round(totalWords * 1.3);
        estimatedTokens = `~${this.formatNumber(tokens)} tokens (~${this.formatNumber(row.classified)} classifications)`;
      } catch { /* ignore */ }
    }

    const lines = [
      "Bot Info:",
      "",
      `Uptime: ${uptime}`,
      `Messages processed (session): ${this.formatNumber(messagesProcessed)}`,
      `Total room messages (all time): ${this.formatNumber(totalRoomMessages)}`,
      `Database size: ${dbSize}`,
      `Active reminders: ${activeReminders}`,
      `LLM classifier: ${llmStatus}`,
      `LLM tokens processed: ${estimatedTokens}`,
    ];

    await this.sendMessage(ctx.roomId, lines.join("\n"));
  }

  private formatUptime(ms: number): string {
    const s = Math.floor(ms / 1000);
    const days = Math.floor(s / 86400);
    const hours = Math.floor((s % 86400) / 3600);
    const minutes = Math.floor((s % 3600) / 60);
    const parts: string[] = [];
    if (days > 0) parts.push(`${days}d`);
    if (hours > 0) parts.push(`${hours}h`);
    parts.push(`${minutes}m`);
    return parts.join(" ");
  }

  private formatBytes(bytes: number): string {
    if (bytes < 1024) return `${bytes} B`;
    if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
    return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
  }

  private formatNumber(n: number): string {
    return n.toLocaleString("en-US");
  }
}

File: internal/plugin/botinfo.go
package plugin

import (
	"context"
	"fmt"
	"log/slog"
	"strings"
	"sync/atomic"
	"time"

	"gogobee/internal/db"
	"gogobee/internal/version"

	"maunium.net/go/mautrix"
)

var (
	botStartTime  = time.Now().UTC()
	botMsgCounter atomic.Int64
)

// IncrementMessageCount increments the global session message counter.
func IncrementMessageCount() {
	botMsgCounter.Add(1)
}

// BotInfoPlugin provides admin-only bot diagnostics.
type BotInfoPlugin struct {
	Base
}

// NewBotInfoPlugin creates a new botinfo plugin.
func NewBotInfoPlugin(client *mautrix.Client) *BotInfoPlugin {
	return &BotInfoPlugin{
		Base: NewBase(client),
	}
}

func (p *BotInfoPlugin) Name() string { return "botinfo" }

func (p *BotInfoPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "botinfo", Description: "Show bot diagnostics (admin only)", Usage: "!botinfo", Category: "Admin", AdminOnly: true},
		{Name: "version", Description: "Show bot version", Usage: "!version", Category: "Info"},
	}
}

func (p *BotInfoPlugin) Init() error { return nil }

func (p *BotInfoPlugin) OnReaction(_ ReactionContext) error { return nil }

func (p *BotInfoPlugin) OnMessage(ctx MessageContext) error {
	if p.IsCommand(ctx.Body, "version") {
		return p.handleVersion(ctx)
	}
	if !p.IsCommand(ctx.Body, "botinfo") {
		IncrementMessageCount()
		return nil
	}

	if !p.IsAdmin(ctx.Sender) {
		return p.SendReply(ctx.RoomID, ctx.EventID, "This command is admin-only.")
	}

	return p.handleBotInfo(ctx)
}

func (p *BotInfoPlugin) handleVersion(ctx MessageContext) error {
	uptime := time.Since(botStartTime)
	days := int(uptime.Hours() / 24)
	hours := int(uptime.Hours()) % 24

	msg := fmt.Sprintf("**%s**\nUptime: %dd %dh", version.Full(), days, hours)

	crashes := db.CrashCount(version.Short())
	if crashes > 0 {
		msg += fmt.Sprintf(" · %d crash(es) this version", crashes)
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}

func (p *BotInfoPlugin) handleBotInfo(ctx MessageContext) error {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("**Bot Diagnostics** — %s\n\n", version.Short()))

	// Uptime
	uptime := time.Since(botStartTime)
	days := int(uptime.Hours() / 24)
	hours := int(uptime.Hours()) % 24
	minutes := int(uptime.Minutes()) % 60
	sb.WriteString(fmt.Sprintf("Uptime: %dd %dh %dm\n", days, hours, minutes))

	// Session message count
	sessionMsgs := botMsgCounter.Load()
	sb.WriteString(fmt.Sprintf("Messages this session: %d\n", sessionMsgs))

	// Total room messages from DB
	d := db.Get()
	var totalMsgs int
	err := d.QueryRow(`SELECT COALESCE(SUM(total_messages), 0) FROM user_stats`).Scan(&totalMsgs)
	if err != nil {
		slog.Error("botinfo: total messages", "err", err)
	}
	sb.WriteString(fmt.Sprintf("Total room messages (DB): %s\n", formatNumber(totalMsgs)))

	// DB size
	var pageCount, pageSize int
	_ = d.QueryRow(`PRAGMA page_count`).Scan(&pageCount)
	_ = d.QueryRow(`PRAGMA page_size`).Scan(&pageSize)
	dbSizeBytes := int64(pageCount) * int64(pageSize)
	dbSizeMB := float64(dbSizeBytes) / (1024 * 1024)
	sb.WriteString(fmt.Sprintf("Database size: %.2f MB\n", dbSizeMB))

	// Active reminders
	var activeReminders int
	_ = d.QueryRow(`SELECT COUNT(*) FROM reminders WHERE fired = 0`).Scan(&activeReminders)
	sb.WriteString(fmt.Sprintf("Active reminders: %d\n", activeReminders))

	// LLM status
	if llmConfigured() {
		sb.WriteString(fmt.Sprintf("LLM status: %s\n", p.checkLLMStatus()))
	} else {
		sb.WriteString("LLM status: not configured\n")
	}

	// Crash stats
	crashesThisVersion := db.CrashCount(version.Short())
	crashesTotal := db.CrashCountAll()
	sb.WriteString(fmt.Sprintf("\nCrashes: %d this version / %d total\n", crashesThisVersion, crashesTotal))

	// Recent crashes (last 5)
	d2 := db.Get()
	rows, err := d2.Query(`SELECT plugin, message, created_at FROM crash_log ORDER BY id DESC LIMIT 5`)
	if err == nil {
		defer rows.Close()
		var hasCrashes bool
		for rows.Next() {
			var plug, msg, ts string
			if rows.Scan(&plug, &msg, &ts) != nil {
				continue
			}
			if !hasCrashes {
				sb.WriteString("Recent crashes:\n")
				hasCrashes = true
			}
			if len(msg) > 80 {
				msg = msg[:80] + "..."
			}
			sb.WriteString(fmt.Sprintf("  %s [%s] %s\n", ts, plug, msg))
		}
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

// checkLLMStatus reports backend liveness for /botinfo. The endpoint it probes
// differs per backend, which the llm package hides behind Ping.
func (p *BotInfoPlugin) checkLLMStatus() string {
	c := llmClient()
	models, err := c.Ping(context.Background())
	if err != nil {
		return fmt.Sprintf("offline (%s: %s)", c.Backend(), err.Error())
	}
	if len(models) == 0 {
		return fmt.Sprintf("online (%s, no models loaded)", c.Backend())
	}
	return fmt.Sprintf("online (%s, %d models: %s)", c.Backend(), len(models), strings.Join(models, ", "))
}
